src/libkernel/sscp/amdgpu/shuffle.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * This file is part of AdaptiveCpp, an implementation of SYCL and C++ standard
+ * parallelism for CPUs and GPUs.
+ *
+ * Copyright The AdaptiveCpp Contributors
+ *
+ * AdaptiveCpp is released under the BSD 2-Clause "Simplified" License.
+ * See file LICENSE in the project root for full license details.
+ */
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "hipSYCL/sycl/libkernel/sscp/builtins/shuffle.hpp"
+#include "hipSYCL/sycl/libkernel/sscp/builtins/detail/shuffle.hpp"
+#include "hipSYCL/sycl/libkernel/sscp/builtins/subgroup.hpp"
+
+template <>
+__acpp_int8 hipsycl::libkernel::sscp::sg_select<__acpp_int8>(__acpp_int8 value, __acpp_int32 id) {
+  return __acpp_sscp_sub_group_select_i8(value, id);
+}
+
+template <>
+__acpp_int16 hipsycl::libkernel::sscp::sg_select<__acpp_int16>(__acpp_int16 value, __acpp_int32 id) {
+  return __acpp_sscp_sub_group_select_i16(value, id);
+}
+
+template <>
+__acpp_int32 hipsycl::libkernel::sscp::sg_select<__acpp_int32>(__acpp_int32 value, __acpp_int32 id) {
+  return __acpp_sscp_sub_group_select_i32(value, id);
+}
+
+template <>
+__acpp_int64 hipsycl::libkernel::sscp::sg_select<__acpp_int64>(__acpp_int64 value, __acpp_int32 id) {
+  return __acpp_sscp_sub_group_select_i64(value, id);
+}
+
+namespace detail {
+static inline unsigned int __lane_id() {
+  return __builtin_amdgcn_mbcnt_hi(-1, __builtin_amdgcn_mbcnt_lo(-1, 0));
+}
+} // namespace detail
+
+HIPSYCL_SSCP_CONVERGENT_BUILTIN
+__acpp_int8 __acpp_sscp_sub_group_shl_i8(__acpp_int8 value, __acpp_uint32 delta) {
+  return __acpp_sscp_sub_group_shl_i32(value, delta);
+}
+
+HIPSYCL_SSCP_CONVERGENT_BUILTIN
+__acpp_int16 __acpp_sscp_sub_group_shl_i16(__acpp_int16 value, __acpp_uint32 delta) {
+  return __acpp_sscp_sub_group_shl_i32(value, delta);
+}
+
+HIPSYCL_SSCP_CONVERGENT_BUILTIN
+__acpp_int32 __acpp_sscp_sub_group_shl_i32(__acpp_int32 value, __acpp_uint32 delta) {
+  auto sg_size = __acpp_sscp_get_subgroup_max_size();
+  int self = detail::__lane_id();
+  int index = (self + delta);
+  index = (int)((self & (sg_size - 1)) + delta) > sg_size ? self : index;
+
+  return __builtin_amdgcn_ds_bpermute(index << 2, value);
+}
+
+HIPSYCL_SSCP_CONVERGENT_BUILTIN
+__acpp_int64 __acpp_sscp_sub_group_shl_i64(__acpp_int64 value, __acpp_uint32 delta) {
+  int tmp[2];
+  __builtin_memcpy(tmp, &value, sizeof(tmp));
```

Comment:
Have you checked whether this affects registerization? `memcpy` semantics forces the compiler to think in terms of memory addresses, which could prevent registerization. Maybe it's smart enough to figure out that it does not need that here.

Might be better to use bit shifts and bitmasks instead.